pkg/apis/pipeline/v1beta1/matrix_types.go

```diff
@@ -37,6 +38,11 @@ type Matrix struct {
 	// +optional
 	// +listType=atomic
 	Include []MatrixInclude `json:"include,omitempty"`
+
+	// combinations is a list of combinations created based on a given Matrix configurations including
+	// Matrix.Params and Matrix.Include.Params
+	// Combinations is skipped while encoding JSON since they are calculated based on the specified params
+	Combinations Combinations `json:"-"`
```

Comment:
why is combinations added to the struct?

Reply:
such that all combinations are stored along with a given `matrix`. This facilitates calling `FanOut` once and store all the combinations for further reuse instead of calling `FanOut` again or passing `combinations` around.

Any cons of introducing this here?

Reply:
I have the same comment as jerop-- I think it's a bit confusing for calling functions to have to know whether or not `FanOut` has been called yet to populate matrix combinations.

Reply:
I do think it would be less confusing to have Combinations separate from the matrix struct. I see in the reconciler code
```
rpt.PipelineTask.Matrix.FanOut()
matrixCombinations := rpt.PipelineTask.Matrix.Combinations.ToParams()
```

and I think 
```
combinations := rpt.PipelineTask.Matrix.FanOut().ToParams()
```
is clearer. (`FanOut` is not something I'd expect to be mutating, and `rpt.PipelineTask.Matrix.Combinations.ToParams()` would produce the wrong result if FanOut hadn't been called.)